### PR TITLE
4886 Add Group email despatch table and admin

### PR DESCRIPTION
**Pull request by sophiemachin**
Adds the groupemaildespatch and userlistlist tables

### CODE REVIEW

**Comment by this-andy** [comment]
https://this.tpondemand.com/entity/4886

File: thiscovery_admin/projects/admin.py [hunk 1]
```diff
@@ -715,6 +716,15 @@ class UserListAdmin(admin.ModelAdmin):
...
 
+class GroupEmailDespatchAdmin(admin.ModelAdmin):
+    list_display = [
+        "id",
+        "created",
+        "template_id",
+        "project",
+    ]
+
+
```
[the remark this answers is not in the document]
Comment (sophiemachin) [reply thiscovery_admin/projects/admin.py:727]: Do we actually need the `id` in the list? When displaying many columns it's a bit cramped on small screens so I'm tempted to leave it out

File: thiscovery_admin/projects/models.py [hunk 2]
```diff
@@ -477,3 +477,27 @@ class UserListMembership(TimeStampedModel):
...
+
+class GroupEmailDespatch(TimeStampedModel):
+    send_start_date = models.DateTimeField()
+    send_end_date = models.DateTimeField()
+    scheduled_date = models.DateTimeField()
+    project = models.ForeignKey(Project, on_delete=models.CASCADE)
+    # This is the ID used in the DynamoDB table
+    template_id = models.CharField(max_length=50)
+    sender = models.ForeignKey(User, on_delete=models.CASCADE)
+    # TODO do we want a ManyToManyField here? Or leave it up to the user to
```
[the remark this answers is not in the document]
Comment (sophiemachin) [reply thiscovery_admin/projects/models.py:493]: Going with many-to-many

File: thiscovery_admin/projects/admin.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (this-andy) [reply thiscovery_admin/projects/admin.py:728]: Possibly not.  Happy to leave it out.  We can always add later if we find it's useful

**Comment by sophiemachin** [comment]
@this-andy This is ready to be looked at. I haven't done UserEmailDespatch, if you'd rather review the whole lot I'll just keep going and then retag you when I'm done

File: thiscovery_admin/projects/migrations/0084_userlistlist.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        migrations.CreateModel(
+            name='UserListList',
+            fields=[
+                ('id', models.UUIDField(default=uuid.uuid4, editable=False, primary_key=True, serialize=False)),
+                ('created', models.DateTimeField(auto_now_add=True)),
+                ('modified', models.DateTimeField(auto_now=True)),
+                ('aggregated_user_list_id', models.UUIDField(default=uuid.uuid4, editable=False)),
+                ('include', models.BooleanField()),
+                ('group_email_despatch', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='projects.groupemaildespatch')),
+                ('user_list', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='projects.userlist')),
```
Comment (this-andy) [inline thiscovery_admin/projects/migrations/0084_userlistlist.py:24]: My recollection was that we would have one field here - the aggregated user list - that would refer to an id in the aggregated view.  And that therefore we didn't need this FK any more.  Am I misremembering? (which is quite possible!)

File: thiscovery_admin/projects/migrations/0084_userlistlist.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            ],
+            options={
+                'unique_together': {('user_list', 'group_email_despatch')},
```
Comment (this-andy) [inline thiscovery_admin/projects/migrations/0084_userlistlist.py:27]: Should this unique constraint refer to aggregated user list id?

File: thiscovery_admin/projects/models.py [hunk 5]
```diff
@@ -477,3 +477,32 @@ class UserListMembership(TimeStampedModel):
...
+    description = models.CharField(max_length=500, default="")
+
+    def __str__(self):
+        return self.description
+
+
+class UserListList(TimeStampedModel):
+
+    aggregated_user_list_id = models.UUIDField(default=uuid.uuid4, editable=False)
+    user_list = models.ForeignKey(UserList, on_delete=models.CASCADE)
```
Comment (this-andy) [inline thiscovery_admin/projects/models.py:503]: comment as above

File: thiscovery_admin/projects/models.py [hunk 6]
```diff
@@ -477,3 +477,32 @@ class UserListMembership(TimeStampedModel):
...
+    group_email_despatch = models.ForeignKey(GroupEmailDespatch, on_delete=models.CASCADE)
+    include = models.BooleanField()
+
+    class Meta:
```
Comment (this-andy) [inline thiscovery_admin/projects/models.py:507]: comment as above

File: thiscovery_admin/projects/migrations/0084_userlistlist.py (hunk 3 above)
Comment (sophiemachin) [reply thiscovery_admin/projects/migrations/0084_userlistlist.py:24]: Oh I've remembered now, I need to write some custom SQL to get this info into django, not use a FK. I don't think we need to talk through this I'll just add it in.

**Comment by sophiemachin** [comment]
@this-andy I've added UserEmailDespatch. template_id/email_type was ambiguous so I ran with template_id, happy to change that if you want.

File: thiscovery_admin/projects/migrations/0085_useremaildespatch.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    operations = [
+        migrations.CreateModel(
+            name='UserEmailDespatch',
+            fields=[
+                ('id', models.UUIDField(default=uuid.uuid4, editable=False, primary_key=True, serialize=False)),
+                ('created', models.DateTimeField(auto_now_add=True)),
+                ('modified', models.DateTimeField(auto_now=True)),
+                ('template_id', models.CharField(max_length=50)),
+                ('opened', models.BooleanField(default=False)),
```
Comment (this-andy) [inline thiscovery_admin/projects/migrations/0085_useremaildespatch.py:22]: Could we make both these booleans nullable and default null, with T/F meaning we actually have data

Comment (sophiemachin) [reply thiscovery_admin/projects/migrations/0085_useremaildespatch.py]: Done

**Review by this-andy — APPROVED** [review]

**Pull request merged**